keras_cv/datasets/waymo/transformer.py

```diff
@@ -0,0 +1,610 @@
+import tensorflow as tf
+from waymo_open_dataset import dataset_pb2
+from waymo_open_dataset.utils import box_utils
+from waymo_open_dataset.utils import frame_utils
+from waymo_open_dataset.utils import range_image_utils
+from waymo_open_dataset.utils import transform_utils
+
+from keras_cv.datasets.waymo import struct
+
+WOD_FRAME_OUTPUT_SIGNATURE = {
+    "frame_id": tf.TensorSpec((), tf.int64),
+    "timestamp_offset": tf.TensorSpec((), tf.float32),
+    "timestamp_micros": tf.TensorSpec((), tf.int64),
+    "pose": tf.TensorSpec([4, 4], tf.float32),
+    "point_smooth_xyz": tf.TensorSpec([None, 3], tf.float32),
+    "point_feature": tf.TensorSpec([None, 4], tf.float32),
+    "point_mask": tf.TensorSpec([None], tf.bool),
+    "point_ri_row_col_sensor_id": tf.TensorSpec([None, 3], tf.float32),
+    "label_box": tf.TensorSpec([None, 7], tf.float32),
+    "label_box_id": tf.TensorSpec([None], tf.int64),
+    "label_box_meta": tf.TensorSpec([None, 4], tf.float32),
+    "label_box_class": tf.TensorSpec([None], tf.int32),
+    "label_box_density": tf.TensorSpec([None], tf.int32),
+    "label_box_detection_difficulty": tf.TensorSpec([None], tf.int32),
+    "label_box_mask": tf.TensorSpec([None], tf.bool),
+    "label_point_class": tf.TensorSpec([None], tf.int32),
+    "label_point_nlz": tf.TensorSpec([None], tf.int32),
+}
+
+_MAX_NUM_NON_TOP_LIDAR_POINTS = 30000
+
+
+def _decode_range_images(frame: dataset_pb2.Frame) -> Dict[int, List[tf.Tensor]]:
+    """Decodes range images from a Waymo Open Dataset frame.
+
+    Args:
+      frame: a Waymo Open Dataset frame.
+
+    Returns:
+      A dictionary mapping from sensor ID to list of range images ordered by
+        return indices.
+    """
+    range_images = {}
+    for lidar in frame.lasers:
+        range_image_str_tensor = tf.io.decode_compressed(
+            lidar.ri_return1.range_image_compressed, "ZLIB"
+        )
+        ri = dataset_pb2.MatrixFloat()
+        ri.ParseFromString(bytearray(range_image_str_tensor.numpy()))
+        ri_tensor = tf.reshape(
+            tf.convert_to_tensor(value=ri.data, dtype=tf.float32), ri.shape.dims
+        )
+        range_images[lidar.name] = [ri_tensor]
+
+        if lidar.name == dataset_pb2.LaserName.TOP:
+            range_image_str_tensor = tf.io.decode_compressed(
+                lidar.ri_return2.range_image_compressed, "ZLIB"
+            )
+            ri = dataset_pb2.MatrixFloat()
+            ri.ParseFromString(bytearray(range_image_str_tensor.numpy()))
+            ri_tensor = tf.reshape(
+                tf.convert_to_tensor(value=ri.data, dtype=tf.float32), ri.shape.dims
+            )
+            range_images[lidar.name].append(ri_tensor)
+    return range_images
+
+
+def _get_point_top_lidar(
+    ris: Sequence[tf.Tensor], frame: dataset_pb2.Frame
+) -> struct.PointTensors:
+    """Gets point related tensors for the top lidar.
+
+    Args:
+      ris: range image tensors. Must have 2 returns. The range image is: [range,
```

Comment:
Why must we have exactly 2 returns?
Maybe since `2` is used as a magic number a few times we should define `_NUM_RETURNS=2` or similar?

Reply:
Yes it is maximum 2 returns.